pkg/client/add_events.go

```diff
@@ -87,6 +96,26 @@ func (client *DataSetClient) AddEvents(bundles []*add_events.EventBundle) error
 	return nil
 }
 
+func (client *DataSetClient) fixServerHostsInBundles(bundles []*add_events.EventBundle) {
+	for _, bundle := range bundles {
+		_, hasOrig := bundle.Event.Attrs[AttrOrigServerHost]
+
+		if hasOrig {
+			// if the orig is set, we are done
+			continue
+		}
+		host, hasHost := bundle.Event.Attrs[AttrServerHost]
+		if hasHost && len(host.(string)) > 0 {
+			// host is set, so lets set is orig as well
+			bundle.Event.Attrs[AttrOrigServerHost] = host
+			continue
+		}
+
+		// as fallback use the value set to the client
+		bundle.Event.Attrs[AttrOrigServerHost] = client.serverHost
```

Comment:
By default, if all the events in a batch (bundle which is part of the same outgoing addEvents HTTP request) contain the same serverHost, we don't need to set it on the event level - we can just set it as part of session / server side attributes. This way we avoid the overhead of setting it on the event level.

Setting it on the event level is only needed in case there is at least one event in the batch which contains ``__origServerHost`` - in that case we want to set ``__origServerHost`` on every single event in the batch and not set it as part of session / server side fields.

It's also worth noting that this implementation right now is redundant / may also be confusing in the UI - if serverHost is present on the event, we just want to set ``_origServerHost``, but not also ``serverHost`` attribute. Here is an example of how we do that in logstash - https://github.com/scalyr/logstash-output-scalyr/blob/master/lib/logstash/outputs/scalyr.rb#L847

Reply:
Yes, I am aware, that the logStash is removing `serverHost` attribute. It's missing here, since I am not sure, whether it is good idea to remove some attributes, which may be set by the user.


And for the logic, the full plan is:
1. Since the batch is mix, we have to populate `__origServerHost` so that it's properly grouped and assigned to the correct session. <- this is done in this PR
2. once the events are in the proper sessions - we can move the value from the `__origServerHost` to the session level.
3. Once it's specified on the session level, we can remove it from the events.

In the log stash code - https://github.com/scalyr/logstash-output-scalyr/blob/master/lib/scalyr/common/util.rb#L131 - there is this function, that fills in the serverHost for all events.

Reply:
When I check the code:
* create_multi_event_request - https://github.com/scalyr/logstash-output-scalyr/blob/master/lib/logstash/outputs/scalyr.rb#L1071 - has some logic for seeing the `serverHost` on the session level
* set_session_level_serverhost_on_events - https://github.com/scalyr/logstash-output-scalyr/blob/master/lib/scalyr/common/util.rb#L131

I would say that the algorithm that I have described above will achieve the same stuff. After that algorithm is used, then the following things are true:
* all events in the same buffer belongs have the same serverHost
* therefore attributes `serverHost` and `__origServerHost` could be removed
* therefore only the `sessionInfo.serverHost` attribute can be filled in

Then it would be working as intended. Is this correct?

Reply:
From your description it sounds correct.

As far as "therefore attributes serverHost and __origServerHost could be removed" goes, that's not implemented yet, right? I assume this would also happen inside dataset-go? If so when / if do we plan to implement it?

As far as this goes:

> It's missing here, since I am not sure, whether it is good idea to remove some attributes, which may be set by the user.

If we don't remove it, it will cause confusion to the end user and possible also issue / undefined behavior on the backend (but for that, I would need to try and reproduce this scenario).

Reply:
Now it's implement it, now I will try to implement tests that will prove, that it works as expected.